test/src/testkernel.ts

```diff
@@ -1245,6 +1246,488 @@ describe('jupyter.services - kernel', () => {
+              expect(calls).to.eql(['last', 'iopub', 'first', 'last', 'iopub']);
+              doLater(() => {
+                done();
+              });
+            };
+          });
+        });
+      });
+
+      it('removing a message hook in another message hook deactivates it immediately', (done) => {
+        let tester = new KernelTester();
+        createKernel(tester).then(kernel => {
+          let options: KernelMessage.IExecuteRequest = {
+            code: 'test',
+            silent: false,
+            store_history: true,
+            user_expressions: {},
+            allow_stdin: false,
+            stop_on_error: false
+          };
+          let future = kernel.execute(options, false);
+          tester.onMessage((message) => {
+            // send a reply
+            let parent_header = message.header;
+            let msg = createMsg('shell', parent_header);
+            tester.send(msg);
+
+            future.onReply = () => {
+              // trigger onIOPub with a 'stream' message
+              let msgStream = createMsg('iopub', parent_header);
+              msgStream.header.msg_type = 'stream';
+              msgStream.content = { 'name': 'stdout', 'text': 'foo' };
+              tester.send(msgStream);
+              // trigger onDone
+              let msgDone = createMsg('iopub', parent_header);
+              msgDone.header.msg_type = 'status';
+              (msgDone as KernelMessage.IStatusMsg).content.execution_state = 'idle';
+              tester.send(msgDone);
+            };
+
+            let calls: string[] = [];
+            let toDelete = (msg: KernelMessage.IIOPubMessage) => {
+              calls.push('delete');
+              return true;
+            }
+            let toDeleteHook = kernel.registerMessageHook(parent_header.msg_id, toDelete);
+
+            kernel.registerMessageHook(parent_header.msg_id, (msg) => {
+              if (calls.length > 0) {
+                // delete the hook the second time around
+                toDeleteHook.dispose();
+              }
+              calls.push('first');
+              return true;
+            });
+
+            future.onIOPub = () => {
+              calls.push('iopub')
+            };
+
+            future.onDone = () => {
+              expect(calls).to.eql(['first', 'delete', 'iopub', 'first', 'iopub']);
+              doLater(() => {
+                done();
+              });
+            };
+          });
+        });
+      });
+    });
+  });
+
+  describe('IFuture', () => {
+    it('should have a read-only msg attribute', (done) => {
+      createKernel().then(kernel => {
+        let future = kernel.execute({ code: 'hello' });
+        expect(typeof future.msg.header.msg_id).to.be('string');
+        expect(() => { future.msg = null; }).to.throwError();
+        done();
+      });
+    });
+
+    describe('Message hooks', () => {
+      it('should have the most recently registered hook run first', (done) => {
```

Comment:
Never mind, I now see that the pattern in the file is to put a newline after describe/context blocks.